File: src/main/java/de/fishdevelopments/fishware/mixin/TridentItemMixin.java
package de.fishdevelopments.fishware.mixin;

import de.fishdevelopments.fishware.Fishware;
import de.fishdevelopments.fishware.features.module.impl.exploit.TridentBoostModule;
import de.fishdevelopments.fishware.manager.impl.ModuleManager;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.item.TridentItem;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.injection.At;
import org.spongepowered.asm.mixin.injection.ModifyArgs;
import org.spongepowered.asm.mixin.injection.Redirect;
import org.spongepowered.asm.mixin.injection.invoke.arg.Args;

@Mixin(TridentItem.class)
public class TridentItemMixin {
  @ModifyArgs(
      method = "onStoppedUsing",
      at =
          @At(
              value = "INVOKE",
              target = "Lnet/minecraft/entity/player/PlayerEntity;addVelocity(DDD)V"))
  public void tridentboost_multiplier(Args args) {
    TridentBoostModule tridentBoostModule =
        Fishware.Companion.getINSTANCE()
            .getManagerManager()
            .get(ModuleManager.class)
            .get(TridentBoostModule.class);
    if (tridentBoostModule.getEnabled()) {
      double multiplier = tridentBoostModule.getMultiplierSetting().getValue();
      args.set(0, (double) args.get(0) * multiplier);
      args.set(1, (double) args.get(1) * multiplier);
      args.set(2, (double) args.get(2) * multiplier);
    }
  }

  @Redirect(
      method = {"use", "onStoppedUsing"},
      at =
          @At(
              value = "INVOKE",
              target = "Lnet/minecraft/entity/player/PlayerEntity;isTouchingWaterOrRain()Z"))
  public boolean tridentboost_allowoutofwater(PlayerEntity instance) {
    TridentBoostModule tridentBoostModule =
        Fishware.Companion.getINSTANCE()
            .getManagerManager()
            .get(ModuleManager.class)
            .get(TridentBoostModule.class);
    if (tridentBoostModule.getEnabled()
        && tridentBoostModule.getAllowOutOfWaterSetting().getValue()) {
      return true;
    }
    return instance.isTouchingWaterOrRain();
  }
}

File: src/main/kotlin/de/fishdevelopments/fishware/mixin/TridentItemMixin.kt
package de.fishdevelopments.fishware.mixin

import de.fishdevelopments.fishware.Fishware.Companion.INSTANCE
import de.fishdevelopments.fishware.features.module.impl.exploit.TridentBoostModule
import de.fishdevelopments.fishware.system.manager.impl.ModuleManager
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.item.TridentItem
import org.spongepowered.asm.mixin.Mixin
import org.spongepowered.asm.mixin.injection.At
import org.spongepowered.asm.mixin.injection.ModifyArgs
import org.spongepowered.asm.mixin.injection.Redirect
import org.spongepowered.asm.mixin.injection.invoke.arg.Args

@Mixin(TridentItem::class)
class TridentItemMixin {
  @ModifyArgs(
      method = ["onStoppedUsing"],
      at =
          At(
              value = "INVOKE",
              target = "Lnet/minecraft/entity/player/PlayerEntity;addVelocity(DDD)V"))
  fun tridentboost_multiplier(args: Args) {
    val tridentBoostModule =
        INSTANCE.managerManager.get(ModuleManager::class.java).get(TridentBoostModule::class.java)
    if (tridentBoostModule.enabled) {
      val multiplier = tridentBoostModule.multiplierSetting.value
      args.set(0, args.get(0) as Double * multiplier)
      args.set(1, args.get(1) as Double * multiplier)
      args.set(2, args.get(2) as Double * multiplier)
    }
  }

  @Redirect(
      method = ["use", "onStoppedUsing"],
      at =
          At(
              value = "INVOKE",
              target = "Lnet/minecraft/entity/player/PlayerEntity;isTouchingWaterOrRain()Z"))
  fun tridentboost_allowoutofwater(instance: PlayerEntity): Boolean {
    val tridentBoostModule =
        INSTANCE.managerManager.get(ModuleManager::class.java).get(TridentBoostModule::class.java)
    if (tridentBoostModule.enabled && tridentBoostModule.allowOutOfWaterSetting.value) {
      return true
    }
    return instance.isTouchingWaterOrRain
  }
}
